datafusion-federation/src/lib.rs

```diff
@@ -4,16 +4,50 @@ use std::{
     sync::Arc,
 };
 
-use datafusion::optimizer::analyzer::Analyzer;
+use datafusion::{
+    execution::context::{SessionContext, SessionState},
+    optimizer::{optimizer::Optimizer, OptimizerRule},
+};
 
-mod analyzer;
-pub use analyzer::*;
+mod optimizer;
+pub use optimizer::*;
 mod table_provider;
 pub use table_provider::*;
 
 mod plan_node;
 pub use plan_node::*;
 
+pub fn default_session_state() -> SessionState {
+    let df_state = SessionContext::new().state();
+
+    let rules = default_optimizer_rules();
+    df_state
+        .with_optimizer_rules(rules)
+        .with_query_planner(Arc::new(FederatedQueryPlanner::new()))
+}
+
+pub fn default_optimizer_rules() -> Vec<Arc<dyn OptimizerRule + Send + Sync>> {
+    // Get the default optimizer
+    let df_default = Optimizer::new();
+    let mut default_rules = df_default.rules;
+
+    // Insert the FederationOptimizerRule after the ScalarSubqueryToJoin.
+    // This ensures ScalarSubquery are replaced before we try to federate.
+    let Some(pos) = default_rules
+        .iter()
+        .position(|x| x.name() == "scalar_subquery_to_join")
+    else {
+        panic!("Could not locate ScalarSubqueryToJoin");
+    };
+
+    // TODO: check if we should allow other optimizers to run before the federation rule.
```

Comment:
`ScalarSubqueryToJoin` is fairly early on in the optimizer list. I wonder if we could just add this to the end? `PushDownLimit` and `PushDownFilter` seem beneficial for the federation case.
<img width="779" alt="Screenshot 2024-05-15 at 8 45 36 PM" src="https://github.com/datafusion-contrib/datafusion-federation/assets/879445/ac6b3ccf-d1f4-45d8-8da1-63e77349bdc3">

Reply:
I would suggest that we take the time to think through the implications of having the optimizers run first and land it in separate PRs. Aside from that there are also federation-specific optimizations we can do (hinted at in #23). We may want to make different choices in those than in the local execution case.